core/jobs/batch_jobs/mailchimp_population_jobs.py

```diff
@@ -0,0 +1,188 @@
+        """Base accumulator where the tuples are added."""
+        return []
+
+    def add_input(self, accumulator: List, model: Tuple) -> List:
+        """Append each tuple to the accumulator list."""
+        accumulator.append(model)
+        return accumulator
+
+    def merge_accumulators(self, accumulators: Iterable[List]) -> List:
+        """Merging accumulators is just combining both of them into a single
+        list.
+        """
+        return list(accumulators)
+
+    def extract_output(
+        self, accumulator: List
+    ) -> List:
+        """Output is the accumulator itself."""
+        return accumulator
+
+
+class MailchimpPopulateJob(base_jobs.JobBase):
+    """One-off job for populating the mailchimp db."""
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        # Pcollection that returns the relevant config property with batch
+        # index.
+        config_property = (
+            self.pipeline
+            | 'Get all config properties' >> ndb_io.GetModels(
+                config_models.ConfigPropertyModel.get_all())
+            | 'Get the batch_index_for_mailchimp property value' >> beam.Filter(
+                lambda model: model.id == 'batch_index_for_mailchimp')
+            | 'Convert model into a dict' >> beam.Map(
+                lambda model: (model.id, model.value)))
+        batch_index_dict = beam.pvalue.AsDict(config_property)
+
+        # PCollection with all user ids that have opted in for email
+        # newsletters.
+        relevant_user_ids = (
+            self.pipeline
+            | 'Get all UserEmailPreferencesModel' >> ndb_io.GetModels(
+                user_models.UserEmailPreferencesModel.get_all().order(
+                    'created_on'))
+            | 'Filter for site_updates == true' >> beam.Filter(
+                lambda model: model.site_updates is True)
+            | 'Extract user id as tuple' >> beam.Map(
+                lambda preferences_model: (preferences_model.id, True))
+        )
+
+        # PCollection of all user ids in oppia mapped to their emails.
+        relevant_user_emails = (
+            self.pipeline
+            | 'Get all user settings models' >> ndb_io.GetModels(
+                user_models.UserSettingsModel.get_all())
+            | 'Extract user email' >> beam.Map(
+                lambda settings_model: (settings_model.id, settings_model.email)
+            )
+        )
+
+        # Merge both PCollections to map the user email to the user ids along
+        # with a bool 'True' if they have opted in for newsletter.
+        mailchimp_results = (
+            (relevant_user_emails, relevant_user_ids)
+            | 'Group by user_id' >> beam.CoGroupByKey()
+            | 'Drop user id' >> beam.Values()  # pylint: disable=no-value-for-parameter
+            | 'Filter for valid user emails' >> beam.Filter(
+                lambda email_status: (
+                    len(email_status[1]) == 1 and email_status[1][0] is True)
```

Comment:
Just to check, in what case would the email_status[1][0] not be true? (Since I thought we already filtered out the others in lines 153-154.)

It's fine if this is a defense-in-depth thing but just wanted to make sure I understood. (It might also be worth leaving a comment in the code.)

Reply:
It is never not True, since as you said, I filtered for true initially. 
This is a bit complicated. So, in the first step, I filtered for `true` users, but we don't have email data for anything and `UserSettingsModel` which has email does not have newsletter status values. 

- So, in the first step, I get all user IDs with status `true` (the first filter).
- Then, I get **all** user data.
- Combine both. This results in tuples, where `true` is the last element for users who have status update set to true, but an empty last element for others. So, yes, this is a defense thing since we don't expect any `false` values there. Either `true` or empty.
- So, I filter for this `true` as well, as at this point, we technically have the entire user data once again.

I could have not had the first filter at all, and filtered for `true` just at this step, and it would have worked, but I wasn't sure if all users had a corresponding `UserPreferencesModel`. So, I ended up doing 2 steps of filtering.

Reply:
OK, I'm not against the multi-filtering, but I suggest leaving a comment in the code explaining the above so that the intent is clear.

Reply:
Not relevant anymore, filtering in a more efficient way following @vojtechjelinek's comments.